6677: Search to highlight cells in the flame view

mirage22 opened the pull request:
ticket: https://bugs.openjdk.java.net/browse/JMC-6677
<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
## Progress
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

## Issue
[JMC-6677](https://bugs.openjdk.java.net/browse/JMC-6677): Search to highlight cells in the Flame View


## Approvers
 * Alex Macdonald ([aptmac](@aptmac) - Committer)
 * Marcus Hirt ([hirt](@thegreystone) - **Reviewer**) **Note!** Review applies to 35f3a2ca3ba524a9c5f3feb0a2f506c5891338c4

aptmac commented:
With regards to the width comment on the closed https://github.com/openjdk/jmc/pull/44#issuecomment-580854588,

> I can also totally remove the width scale but I've observed some not wanted behaviour on the MacOS, have you tested it on Mac ?

What was the unwanted behaviour you were seeing?

Here's my initial feedback on the PR:

1. As for the sizing fix, I don't think setting the chart to 95% of the window size [[0]](https://github.com/openjdk/jmc/pull/45/files#diff-30491bb3d763f9e956043f3402596f31R88) is the complete fix; the problem is that the svg doesn't change it's size when the chart is changed. So if the chart increases in size it is cutoff by the svg, and if the chart decreases in size then it leaves excess white space. 

GIF1: Increasing the chart size, but the SVG doesn't change
![6677-increasing-cutoff](https://user-images.githubusercontent.com/10425301/73668357-c1831c80-4673-11ea-93fd-9b6882fd81d6.gif)

GIF2: Decreasing the chart size, but the SVG doesn't change
![6677-decreasing-excess](https://user-images.githubusercontent.com/10425301/73668366-c47e0d00-4673-11ea-87cc-ae3f61f41720.gif)

2. As for the search functionality itself, I think it looks really nice. My feedback for the moment is that it is using `startsWith()` for checking the strings, so it's only possible to search by class .. it would be neat to be able to search within the string as well. e.g., search for `.run()` and have it highlight all instances of `.run()` being called.

[0] https://github.com/openjdk/jmc/pull/45/files#diff-30491bb3d763f9e956043f3402596f31R88
[1] https://github.com/openjdk/jmc/pull/45/files#diff-c309933c5cd238df4805885ae811bae1R128

mirage22 commented:
@aptmac  I like your comments! 
add1.  agree,  I know about the issue with the window. I've been searching about how to fix it properly and it looks they have opened a bug ticket in flameview. They don't properly propagate resize to the base svg. Maybe we should wait until they have it fixed, what do you think?

add2. I've been thinking about it ! So I'll add it too let's see how performance.

aptmac commented:
> @aptmac I like your comments!
> add1. agree, I know about the issue with the window. I've been searching about how to fix it properly and it looks they have opened a bug ticket in flameview. They don't properly propagate resize to the base svg. Maybe we should wait until they have it fixed, what do you think?

I'm the one who opened that issue on the `d3-flame-graph` repo, and I've posted on one of their issues last week so I'm not sure how long it takes to get a reply (or a fix). I think if going the route of waiting for an upstream fix then we should also update to the latest version .. we're using one 1.5 years out of date at the moment. However, this would also complicate the PR for getting this to work in Windows until Chromium support is added to SWT.

In my opinion this is something that stylesheets can address and doesn't require us to rely on external parties.

> ps: when you click on a new thread to display (example ), all is properly resized according to the window size. (my observed behaviour)

Interesting, I'm only seeing a proper resize when the chart is redrawn.

![6677-resize-on-redraw](https://user-images.githubusercontent.com/10425301/73748384-509d3c80-4727-11ea-9141-8253a8f2cb57.gif)

mirage22 commented:
![image](https://user-images.githubusercontent.com/1956942/73792107-f3939c00-47a3-11ea-8e72-4b687a163e58.png)

@aptmac I've fixed the resizing now it works on my Mac perfectly. 
I've also implement your recommendation according to the search (add.2.). It works pretty well and I think ti may sense.

![image](https://user-images.githubusercontent.com/1956942/73792373-6e5cb700-47a4-11ea-82ce-97208a093978.png)

conclusion: 
1. resize works  (on mac as I've no windows easy to access ;)
2. search considers  node and package name. It searches for the strings that contains desired sequence.
3. I've changed the highlighting color to Magenta 

what do you think now ? @thegreystone and @aptmac

mirage22 commented:
/contributor add @aptmac

mirage22 commented:
/contributor add @thegreystone

mirage22 commented:
/contributor add Marcus Hirt <marcus@hirt.se>

mirage22 commented:
/contributor add Alex Macdonald <Alex.Mcdonald@redhat.com>

aptmac commented on application/org.openjdk.jmc.flightrecorder.flameview/src/main/js/flameviewColoring.js:
```diff
@@ -117,9 +117,21 @@ const createHslColorString = function(h,s,l){
 
 const colorCell = function (d) {
-	return colorByPackage(d.data.p);
+	if (textToSearch !== "" && (evaluateSearchElement(d.data.p) || evaluateSearchElement(d.data.n))){
```
formatting: space before `{`

mirage22 replied:
update

aptmac commented on application/org.openjdk.jmc.flightrecorder.flameview/src/main/js/flameviewColoring.js:
```diff
@@ -117,9 +117,21 @@ const createHslColorString = function(h,s,l){
+	}
 };
 
+const evaluateSearchElement = function(text){
```
formatting: space before `{`

aptmac commented on application/org.openjdk.jmc.flightrecorder.flameview/src/main/js/flameviewColoring.js:
```diff
@@ -117,9 +117,21 @@ const createHslColorString = function(h,s,l){
 };
 
+const evaluateSearchElement = function(text){
+	if((text !== undefined && text.toLowerCase().includes(textToSearch))){
```
formatting: space after `if`
formatting: looks like there's an extra pair of brackets on the if-statement
formatting: space before {

aptmac commented on application/org.openjdk.jmc.flightrecorder.flameview/src/main/resources/page.template:
```diff
@@ -10,15 +10,28 @@
 </head>
 
-<body onresize="resizeFlameGraph()">
+<body>
```
out of curiosity, why the change from using `onresize` to adding a listener to the browser?

mirage22 replied:
This was another tricky part, when the swt.browser is created it gets the flameGraph default size. Even when you resize the jmc window the size is not updated. You need to check the event on the Browser object and force the update by as I'm doing with resizeFlameGraph.

aptmac commented on application/org.openjdk.jmc.flightrecorder.flameview/src/main/resources/page.template:
```diff
@@ -43,13 +56,21 @@
 
+		function executeSearch(){
+			const inputText = packageInputField.value;
+			if(textToSearch !== inputText){
```
formatting: space after `if`
formatting: space before `{`

mirage22 replied:
update

aptmac commented on application/org.openjdk.jmc.flightrecorder.flameview/src/main/resources/page.template:
```diff
@@ -10,15 +10,28 @@
 		
+		const packageInputField = document.getElementById("packageNameInput");
+		packageInputField.addEventListener("keyup", function(event) {
+				if(event.keyCode === 13){
```
formatting: space after `if` and before `{`

mirage22 replied:
update

aptmac commented on application/org.openjdk.jmc.flightrecorder.flameview/src/main/resources/page.template:
```diff
@@ -43,13 +56,21 @@
 				.call(flameGraph);
 		}
 
+		function executeSearch(){
```
formatting: space before `{`

mirage22 replied:
maybe I've forgotten to run spotless, I'll do

aptmac commented on application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/common/ImageConstants.java:
```diff
@@ -110,6 +110,8 @@
 	public static final String ICON_IO_BY_PORT = "io_by_port.png"; //$NON-NLS-1$
 	public static final String ICON_IO_BY_HOST_AND_PORT = "io_by_host_and_port.png"; //$NON-NLS-1$
 
+	public static final String ICON_FLAMEVIEW_SEARCH = "magnifyer64.png"; //$NON-NLS-1$
```
`s/magnifyer/magnifier/`

mirage22 replied:
corrected

mirage22 commented:
@aptmac does it behave differently by you ? 
![image](https://user-images.githubusercontent.com/1956942/73887464-e93cd500-486b-11ea-814c-7d3254b350bd.png)

but you may be right correct with the escaping characters, I'll think about your comments tomorrow.

mirage22 commented:
![image](https://user-images.githubusercontent.com/1956942/73889417-eba12e00-486f-11ea-9e60-4ee67cc82260.png)

@aptmac what do you think about the update ? I'm still thinking about this the automatic chart update and the sufficient special characters filtering

mirage22 commented:
Currently the highlighting is activated when our get out of the field or you press the enter

thegreystone commented:
> Additionally, I found some formatting errors in the commits here .. but there's also quite a bit that made it through previously and it'd be nice to clean that up IMO. Do we have any JavaScript formatting guidelines that we're following?

Yep. They are under configuration/ide/eclipse/formatting/formattingjs.xml.

mirage22 commented:
@aptmac I've added the search execution based on the input filed characters length. It looks very nice, I think it's quite intuitive now. 
I've also added possibility to be able to search special character based string like : 'ArrayList.add(Object [..]).'  you have mentioned. 
![image](https://user-images.githubusercontent.com/1956942/73944619-2dbf8380-48f3-11ea-9021-93bbc698817d.png)

aptmac commented:
If I type a single "special character" (`$`, `%`, etc), then all of the nodes get highlighted. This seems to be the case for most of them, minus a few such as `(`, `)`, `+`.
![6677-highlight-all](https://user-images.githubusercontent.com/10425301/73955325-28e9d800-48d1-11ea-8a31-e14d98d5191f.gif)

It appears to only match strings with parenthesis when the entire string matches. For example, if the string is `ArrayList.add(Object)`, then it correctly matches `.add`, and `.add(Object)`, but will not highlight for `add(` or `add(O` etc.
![6677-match-whole](https://user-images.githubusercontent.com/10425301/73955460-68182900-48d1-11ea-8d8b-2340d56e13c2.gif)

The automatic highlighting after typing feels nice. Should it also clear if there's no text in the search box? If the user types in a string to search and then uses backspaces to clear the text, at the moment the labels will remain highlighted.

mirage22 commented:
@aptmac it's good point ! I've added the backspace check. 
according to the issue: ```...add(Object), but will not highlight for add( or add(O etc.```  . I see the issue with flamegraph lib itself. I'm trying to fix it.

mirage22 commented:
@aptmac  and @thegreystone  I've enabled the full special characters search, it should work as expected.  Everyone should be able find  e.g.: `run(` 
I've not noticed any performance issues, I hope it works similar also for you.

see images 
![image](https://user-images.githubusercontent.com/1956942/73984774-bc56f380-4939-11ea-9c32-8df06059f52b.png)

![image](https://user-images.githubusercontent.com/1956942/73984793-c7aa1f00-4939-11ea-8ddf-b045176d0ee4.png)

![image](https://user-images.githubusercontent.com/1956942/73984817-d55fa480-4939-11ea-9d58-e163238dcd56.png)

aptmac commented on application/org.openjdk.jmc.flightrecorder.flameview/src/main/js/flameviewColoring.js:
```diff
@@ -130,7 +133,15 @@ const evaluateSearchElement = function(text) {
 
 const removeSpecialCharacters = function(text) {
-	return text.trim().replace(/[&\/\\#,+()$~%'":*?<>{}]/g, '_').toLowerCase();
+	//return text.trim().replace(/[&\/\\#,+()$~%'":*?<>{}]/g, '_').toLowerCase();
```
leftover comment

aptmac commented on application/org.openjdk.jmc.flightrecorder.flameview/src/main/js/flameviewColoring.js:
```diff
@@ -130,7 +133,15 @@ const evaluateSearchElement = function(text) {
-	return text.trim().replace(/[&\/\\#,+()$~%'":*?<>{}]/g, '_').toLowerCase();
+	//return text.trim().replace(/[&\/\\#,+()$~%'":*?<>{}]/g, '_').toLowerCase();
+	const map = Array.prototype.map;
+    const resultText = map.call(text.trim().toLowerCase(), element => {
```
indent level is off, and the following code block has some spacing to fix up

mirage22 replied:
;) it's not there

aptmac commented on application/org.openjdk.jmc.flightrecorder.flameview/src/main/js/flameviewColoring.js:
```diff
@@ -60,6 +60,9 @@ const packageMarkerRest = "rest";
 const packagesIdentifierMap = new Map().set("java.", packageMarkerJava).set("sun.", packageMarkerSun).set("com.sun.", 
 		packageMarkerComSunAndJdk).set("jdk.", packageMarkerComSunAndJdk);
 const packageColorMap = new Map().set("", rootPackageColor);
+const specialCharactersMap = new Map().set('#','\x23').set('$','\x24').set('(', '\x28').set(')','\x29')
```
inconsistencies in spaces after the `,` between the special character and it's alternate value

mirage22 commented:
/integrate

mirage22 commented:
/integrate

mirage22 commented:
/integrate

aptmac commented on application/org.openjdk.jmc.flightrecorder.flameview/src/main/js/flameviewColoring.js:
```diff
@@ -116,13 +119,33 @@ const createHslColorString = function(h,s,l){
+};
+
+const removeSpecialCharacters = function(text) {
+	const map = Array.prototype.map;
```
if the map isn't used again, why not have the next line be:
`const resultText = Array.prototype.map.call[..]` ?

mirage22 replied:
good point, i've it overlooked

mirage22 commented:
/integrate

mirage22 commented:
/contributor remove Alex Macdonald Alex.Mcdonald@redhat.com

mirage22 commented:
/contributor add Alex Macdonald almacdon@redhat.com

mirage22 commented:
/contributor remove Alex Mcdonald <Alex.Mcdonald@redhat.com>

mirage22 commented:
/contributor add Alex Mcdonald <almacdon@redhat.com>

mirage22 commented:
/integrate

mirage22 commented:
/contributor remove Alex Macdonald <Alex.Mcdonald@redhat.com>

mirage22 commented:
/contributor add Alex Macdonald <almacdon@redhat.com>

mirage22 commented:
/integrate

thegreystone commented:
/sponsor
